Camera replaced by Cinematographic Camera (CinemAirSim)

ppueyor opened the pull request:
<!-- Thank you for submitting a pull request! -->
<!-- ⚠️⚠️ Do Not Delete This! pull_request_template ⚠️⚠️ -->
<!-- Please read our contribution guidelines: https://microsoft.github.io/AirSim/CONTRIBUTING/ -->

Fixes: #1489   <!-- add this line for each issue your PR solves. -->
Fixes: #2303
Fixes: #2311
-Integrates the CineCameraComponent into AirSim
-New PR from https://github.com/microsoft/AirSim/pull/3779

## About
An AddOn for AirSim that includes all the tools to integrate a cinematographic camera and the needed tools to control/access to it.
More info available in repo: https://github.com/ppueyor/CinemAirSim
Or conference paper: https://arxiv.org/abs/2003.07664

## How Has This Been Tested?
Ubuntu 20 and UE4 4.26
Different demos and API calls
Results available in repo

## Screenshots:
https://www.youtube.com/watch?v=OJGNaitWZVA

Commit:
Camera replaced by Cinematographic Camera (CinemAirSim)

Commit:
Camera replaced by Cinematographic Camera (CinemAirSim)

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,188 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+    return ss.str();
+}
+
+void APIPCamera::CopyCameraSettingsToAllSceneCapture(UCameraComponent* camera_)
```
According to the [naming conventions](https://microsoft.github.io/AirSim/coding_guidelines/#naming-conventions), the underscore suffix is used for member variables, this should just be `camera`: this is also causing a build error on my side:

C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(750): error C4458: declaration of 'camera_' hides class member
  C:\__w\13\s\Unreal\Environments\Blocks\Plugins\AirSim\Source\PIPCamera.h(100): note: see declaration of 'APIPCamera::camera_'

zimmy87 replied:
this has been addressed in latest iteration

Commit:
Added methods to Python Client, fixed bug when packaging

Commit:
fixed indentation bug in Python client

rajat2004 commented on AirLib/include/api/RpcLibClientBase.hpp:
```diff
@@ -111,6 +111,24 @@ namespace airlib
+        //CinemAirSim
+        std::vector<std::string> simGetPresetLensSettings(const std::string& camera_name, const std::string& vehicle_name = "", bool external = false);
+        std::string simGetLensSettings(const std::string& camera_name, const std::string& vehicle_name = "", bool external = false);
+        void simSetPresetLensSettings(const std::string& preset_lens_settings = "", const std::string& camera_name = "", const std::string& vehicle_name = "", bool external = false);
```
`preset_lens_settings` and `camera_name` defaults probably shouldn't be present

zimmy87 replied:
this has been addressed in latest revision

rajat2004 commented on AirLib/include/api/RpcLibClientBase.hpp:
```diff
@@ -111,6 +111,24 @@ namespace airlib
+        std::string simGetLensSettings(const std::string& camera_name, const std::string& vehicle_name = "", bool external = false);
+        void simSetPresetLensSettings(const std::string& preset_lens_settings = "", const std::string& camera_name = "", const std::string& vehicle_name = "", bool external = false);
+        std::vector<std::string> simGetPresetFilmbackSettings(const std::string& camera_name, const std::string& vehicle_name = "", bool external = false);
+        void simSetPresetFilmbackSettings(const std::string& preset_filmback_settings = "", const std::string& camera_name = "", const std::string& vehicle_name = "", bool external = false);
```
Same, first 2 defaults shouldn't be there

rajat2004 commented on AirLib/src/api/RpcLibClientBase.cpp:
```diff
@@ -271,6 +271,106 @@ __pragma(warning(disable : 4239))
+        {
+            vector<std::string> result = pimpl_->client.call("simGetPresetLensSettings", camera_name, vehicle_name, external).as<vector<std::string>>();
+
+            if (result.size() == 1) {
+                // rpclib has a bug with serializing empty vectors, so we return a 1 byte vector instead.
+                result.clear();
+            }
```
No longer needed, can be removed from other places also

rajat2004 commented on AirLib/src/api/RpcLibServerBase.cpp:
```diff
@@ -153,6 +153,86 @@ namespace airlib
+        });
+
+        pimpl_->server.bind("simGetLensSettings", [&](const std::string& camera_name, const std::string& vehicle_name, bool external) -> string {
+            auto result = getWorldSimApi()->getLensSettings(CameraDetails(camera_name, vehicle_name, external));
+            return result;
```
Just directly return instead of creating a var, can be cleaned up in other places after removing the vector checking also

rajat2004 commented on PythonClient/airsim/client.py:
```diff
@@ -265,7 +265,74 @@ def simGetImages(self, requests, vehicle_name = '', external = False):
+#CinemAirSim
+    def simGetPresetLensSettings(self, camera_name, vehicle_name = '', external = False):  
+        result = self.client.call('simGetPresetLensSettings', camera_name, vehicle_name, external)
+        if (result == "" or result == "\0"):
+            return None
```
Again shouldn't be needed I think, but not 100% sure about this. There's no such handling inside current master I think

ppueyor replied:
This is from the master branch!

rajat2004 reviewed (commented):
The video and paper looks pretty awesome, and it's pretty nice that you want this to get merged in the project also
Just had a brief look, and had some concerns hope you might be able to answer, and might interest the maintainers also
- What's the performance diff if any between the current camera and the new one? Some FPS info would be good, #3889 might be useful. Also, does this support all the image types?
- Wondering whether replacing the current camera entirely might be a good idea or not. This adds a lot of camera APIs, and there's enough already. In case of problems with first point, whether having this as a separate camera might be a better option. A separate set of APIs (client) for camera might be something with looking into, it'll improve the current code also by avoiding mixing so many APIs

ppueyor commented:
Hi @rajat2004,

Thanks for your interest in this PR. My answers are below:

- I used that script and the average FPS is 19.90, don't know how much it is in the regular camera. It does support all kinds of image types, and it is possible to get images and subwindows the same way and updates as the regular camera.

- Well, I think that it's something that the maintainers should answer, if that could be interesting or not. I can talk about my own experience. It opened me to a whole world of possibilities. If you want to use the drone with filming objectives (that it is wide use of the UAVs), this PR is more than helpful because u could not modify the intrinsics of the camera otherwise.

About the corrections, @zimmy87 is having a look at the code, so if he considers that, I can correct the code with your suggestions.

rajat2004 commented:
By the FPS, I meant some results to compare on your system itself, for the Cinematic and normal cameras, for different resolutions like 256x144, 720p, 1080p and maybe 4k also?
Makes sense, UAVs for filming is probably the most common usecase, other applications are also closely vision-related, and improvements in camera to model the real world would definitely be useful

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,188 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+    return ss.str();
+}
+
+void APIPCamera::CopyCameraSettingsToAllSceneCapture(UCameraComponent* camera)
```
function names should be camelCase; this applies to APIPCamera::CopyCameraSettingsToSceneCapture as well

ppueyor replied:
True!

zimmy87 replied:
fixed in latest revision

zimmy87 reviewed (changes requested):
Thanks for submitting this PR, this is a cool feature! Most of my feedback is style-related. I second @rajat2004's comment about replacing the default camera with the new UCineCameraComponent. I think the default user experience is different enough that there should be an option for switching to the new type of camera. It'd also be nice to see the rest of @rajat2004's review items addressed before merging this.

zimmy87 commented:
Our Azure pipeline check is failing with the following build error:

```

In file included from /__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Intermediate/Build/Linux/B4D820EA/UE4/Development/AirSim/Module.AirSim.2_of_2.cpp:2:
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/TextureShuffleActor.cpp:24:49: error: incomplete type 'UMaterialInstanceDynamic' named in nested name specifier
        DynamicMaterialInstances[material_id] = UMaterialInstanceDynamic::Create(DynamicMaterial, this);
                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ue4/ue-4.25.1-linux-debugeditor/Engine/Source/Runtime/Engine/Classes/Engine/BlendableInterface.h:10:7: note: forward declaration of 'UMaterialInstanceDynamic'
class UMaterialInstanceDynamic;
      ^
In file included from /__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Intermediate/Build/Linux/B4D820EA/UE4/Development/AirSim/Module.AirSim.2_of_2.cpp:2:
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/TextureShuffleActor.cpp:27:33: error: member access into incomplete type 'UStaticMeshComponent'
        components[component_id]->SetMaterial(material_id, DynamicMaterialInstances[material_id]);
                                ^
/home/ue4/ue-4.25.1-linux-debugeditor/Engine/Source/Runtime/CoreUObject/Public/Templates/Casts.h:22:7: note: forward declaration of 'UStaticMeshComponent'
class UStaticMeshComponent;
      ^
In file included from /__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Intermediate/Build/Linux/B4D820EA/UE4/Development/AirSim/Module.AirSim.2_of_2.cpp:2:
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/TextureShuffleActor.cpp:30:42: error: member access into incomplete type 'UMaterialInstanceDynamic'
    DynamicMaterialInstances[material_id]->SetTextureParameterValue("TextureParameter", SwappableTextures[tex_id]);
                                         ^
/home/ue4/ue-4.25.1-linux-debugeditor/Engine/Source/Runtime/Engine/Classes/Engine/BlendableInterface.h:10:7: note: forward declaration of 'UMaterialInstanceDynamic'
class UMaterialInstanceDynamic;
      ^
4 errors generated.
```

Are you seeing this in local builds?

ppueyor commented on PythonClient/airsim/client.py:
```diff
@@ -16,7 +16,7 @@ def __init__(self, ip = "", port = 41451, timeout_value = 3600):
         self.client = msgpackrpc.Client(msgpackrpc.Address(ip, port), timeout = timeout_value, pack_encoding = 'utf-8', unpack_encoding = 'utf-8')
 
-    # -----------------------------------  Common vehicle APIs ---------------------------------------------
+#-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- - Common vehicle APIs -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -
```
I did not add those spaced, it should be something related to the clang script, I didn't even know they were there. We can delete them of course.

zimmy87 replied:
Our .clang-format configuration file only currently supports cpp, it shouldn't be reformatting python code. Are you by chance using a configuration defined elsewhere? Looks like this has been partially fixed in the latest revision, but I'm still seeing some changes in whitespace. I would like to see these changes in whitespace reverted before merging this.

ppueyor replied:
Okay, I was checking all the modified files. Updated!

ppueyor commented on Unreal/Plugins/AirSim/Source/PIPCamera.h:
```diff
@@ -69,8 +95,9 @@ class AIRSIM_API APIPCamera : public ACameraActor
+    //CinemAirSim
     UPROPERTY()
-    UCameraComponent* camera_;
+    UCineCameraComponent* camera_;
```
What do you think is the best option to resolve this? Maybe a new entry in the settings.json could enable/disable all the cinematographic features. The thing is that some of the calls to the RPCLib are just enabled with this kind of camera. Is there an option to avoid these functions to be called if the embedded camera is not cinematographic?
Programatically, it would be very easy to implement. When the PIPCamera is embedded in the drones, we can introduce an if to introduce one or another kind of PIPCamera.

zimmy87 replied:
Personally, I think the best option here would be to introduce a new boolean setting inside both the CameraDefaults element and in https://microsoft.github.io/AirSim/settings/#camera-settings. This new boolean setting could be stored inside each APIPCamera instance. For CinemAirSim-specific methods, I think it would be best to have these essentially be no-ops whenever the setting is set to a regular UCameraComponent. To communicate to the user that these methods aren't doing anything, we could output a log statement saying the method being run doesn't perform any operation on a UCameraComponent. Another possibility is to throw an exception, but personally I don't think this warrants stopping execution of the program, just messaging to the user that the methods aren't doing anything should be sufficient.

ppueyor replied:
The problem I see here is that the APIPCamera is an instance of the method UCineCameraComponent, and the header should be a child of: `class AIRSIM_API APIPCamera : public ACineCameraActor `. It is something that depends on the whole class, not only on one attribute of the class that we can select/deselect.

We can eventually have two separate classes, PIPCamera and PIPCinemaCamera, for instance, and attach to the drone one or another depending on the boolean set in the settings, when these cameras are created. What do you think?

zimmy87 replied:
In that case, would it be possible to change the default settings of the UCineCameraComponent property to resemble those of a default UCameraComponent? We're not concerned about introducing a dependency on UCineCameraComponent or ACineCameraActor, it's moreso we just think it'll be confusing for users to see a camera with a different field of view or other characteristics than before.

ppueyor replied:
I think that dependency is needed, so the camera on the actor is a CineCameraComponent instead of a CameraComponent, with all the fields and parameters that are only of that class, and in the way it generates a BluePrint with a CineCamera instead of a regular Camera.


What about two different PIPCameras? One that depends on CameraActor and another that depends on CineCameraActor.

ppueyor commented:
Hi @zimmy87, I answered your suggestions/bug reports. I will fix everything once we decide what to do to integrate it.

No, I haven't seen that error of compilation in local builds. I'm using UE 4.26.1 in Ubuntu 20. Can this be the difference?

What are the next steps we should do? Should I wait for your instructions?

Commit:
Fixed naming conventions and minor bugs requested in the PR

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,188 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+
+        ss << "Name: " << name << ";\n\t MinFocalLength: " << preset.LensSettings.MinFocalLength << "; \t MaxFocalLength: " << preset.LensSettings.MaxFocalLength;
+        ss << "\n\t Min FStop: " << preset.LensSettings.MinFStop << "; \t Max Fstop: " << preset.LensSettings.MaxFStop;
+        //ss << "\n\t Diaphragm Blade Count: " << preset.LensSettings.DiaphragmBladeCount;
```
fixed in latest revision

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,186 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+void APIPCamera::copyCameraSettingsToAllSceneCapture(UCameraComponent* camera)
+{
+    int image_count = static_cast<int>(Utils::toNumeric(ImageType::Count));
+    for (image_type >= 0) {
```
Have you tested this method? Unless I'm missing something, I think image_type needs to be initialized and  incremented/decremented. I've never seen a for loop with just a condition before.

zimmy87 replied:
This is causing build errors in our Azure Pipelines on Windows:

```
C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(752): error C2065: 'image_type': undeclared identifier
  C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(752): error C2143: syntax error: missing ';' before ')'
  C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(752): error C2143: syntax error: missing ';' before '{'
  C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(752): error C2143: syntax error: missing ')' before '{'
  C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(754): error C2065: 'image_type': undeclared identifier
  C:/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp(754): error C2660: 'APIPCamera::copyCameraSettingsToSceneCapture': function does not take 1 arguments
  C:\__w\13\s\Unreal\Environments\Blocks\Plugins\AirSim\Source\PIPCamera.h(65): note: see declaration of 'APIPCamera::copyCameraSettingsToSceneCapture'
```

as well as on Ubuntu:

```
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp:752:10: error: use of undeclared identifier 'image_type'
    for (image_type >= 0) {
         ^
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp:752:25: error: expected ';' in 'for' statement specifier
    for (image_type >= 0) {
                        ^
/__w/13/s/Unreal/Environments/Blocks/Plugins/AirSim/Source/PIPCamera.cpp:754:61: error: use of undeclared identifier 'image_type'
        copyCameraSettingsToSceneCapture(camera_, captures_[image_type]);
                                                            ^
```

ppueyor replied:
My mistake, some lines should remain out of the code in the last commit, sorry for this. 

In my local version this method looks like this:

``` 
void APIPCamera::copyCameraSettingsToAllSceneCapture(UCameraComponent* camera)
{
   int image_count = static_cast<int>(Utils::toNumeric(ImageType::Count));
    for (int image_type = image_count - 1; image_type >= 0; image_type--) {
        if (image_type >= 0) { //scene capture components
            CopyCameraSettingsToSceneCapture(camera_, captures_[image_type]);
        }
    }
}
```

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,188 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+void APIPCamera::CopyCameraSettingsToAllSceneCapture(UCameraComponent* camera)
+{
+    int image_count = static_cast<int>(Utils::toNumeric(ImageType::Count));
+    for (int image_type = image_count - 1; image_type >= 0; image_type--) {
```
this comment no longer applies to the latest revision

ppueyor replied:
Got it! You're absolutely right.

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -588,3 +595,188 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+        dst->SetWorldLocationAndRotation(src->GetComponentLocation(), src->GetComponentRotation());
+        dst->FOVAngle = src->FieldOfView;
+
+        FMinimalViewInfo CameraViewInfo;
```
camera_view_info has been fixed, but there's still some remaining locals in this method that need to be fixed: namely, SrcPPSettings, DstPPSettings, and DstWeightedBlendables

ppueyor replied:
you're right, sorry for this

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.h:
```diff
@@ -39,6 +43,28 @@ class AIRSIM_API APIPCamera : public ACameraActor
+    void setPresetFilmbackSettings(std::string preset_string);
+    std::string getLensSettings();
+    std::string getFilmbackSettings();
+    float setFilmbackSettings(float sensor_width, float sensot_height);
```
fixed in latest revision

zimmy87 reviewed (changes requested):
Additional comment regarding a build failure

zimmy87 commented:
Hi @ppueyor, apologies for the delay in getting back to you; I've responded to the remaining feedback items; please let me know if you have any other questions/concerns. Regarding the build failure in TextureShuffleActor.cpp, I'm still seeing this with your latest revision, so this looks like a legitimate build failure. My guess is you're missing an include somewhere (usually the "incomplete type" error occurs when only a forward declaration is present for a given class and the full header for that class is needed to access one of its members). We're currently using 4.25.1 on our build machines, and we'd like to maintain compatibility with 4.25, so using 4.26.1 could be a potential cause here. Would you be able to test locally with a 4.25.1 build?

ppueyor commented:
Hi @zimmy87, 

Thanks for the review. Yes, I will test it in Unreal 4.25.1 once we decide what to do with the cinematic flag in Settings.hpp.
I fixed the suggested changes. I will commit everything when it is all decided and ready.

Commit:
added files

Commit:
updated

Commit:
Merge branch 'microsoft-master2' into cinemairsim1

Commit:
Focal length set to match with original PIPCamera

Commit:
Focal length set to match with original PIPCamera

ppueyor commented:
Hi @zimmy87, 

I implemented all the requested changes and updated my repo to the latest version of your master.
I talked to Jonny and we decided to set the focal length to match with the regular camera. 

I tested it with ubuntu 20 and Unreal 4.25 and works perfectly. 

Please, tell me if you need anything else.

Regards,
Pablo

zimmy87 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -370,6 +374,9 @@ void APIPCamera::setupCameraFromSettings(const APIPCamera::CameraSetting& camera
         this->SetActorTickEnabled(false);
 
+    //set initial focal length
+    camera_->CurrentFocalLength = 11.9;
```
Can we move this to APIPCamera::PostInitializeComponents to ensure it happens for all PIPCamera instances, not just the ones attached to the vehicle pawn?

Currently, APIPCamera::setupCameraFromSettings is called from ASimModeBase::createVehicleApi on all cameras attached to the vehicle pawn, so I see the focal length get updated for all those cameras. But there's another camera attached to the CameraDirector that gets created in ASimModeBase::initializeCameraDirector and this camera ends up being the one used for the main game view. Since APIPCamera::setupCameraFromSettings never gets called on this camera, I still see the default camera have the default focal length of 35.

ppueyor replied:
Moved to APIPCamera::PostInitializeComponents. Can you test if it works in that particular case?

zimmy87 reviewed (changes requested):
The new focal length looks good when I manually override the right camera, but unfortunately I'm not seeing this focal length applied to the default camera, so I still see the same narrow field of view with the latest iteration.

zimmy87 commented:
Hi @ppueyor, thanks for the update, I tested locally and noticed a few things that I think should get addressed before merging. I've gone through all previous review feedback and resolved everything that looks to have been addressed. I see that there's one remaining item of feedback from @rajat2004 on PythonClient/airsim/client.py that hasn't been addressed yet (it looks like this may not require a code change, feel free to resolve the comment yourself if that's the case). There also looks like there's one remaining item of feedback from my previous review about the names of SrcPPSettings, DstPPSettings, and DstWeightedBlendables in Unreal/Plugins/AirSim/Source/PIPCamera.cpp. These variables should be renamed to src_pp_settings, dst_pp_settings, and dst_weighted_blendables.

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+        std::ostringstream ss;
+        std::string name = (TCHAR_TO_UTF8(*preset.Name));
+
+        ss << "Name: " << name << ";\n\t MinFocalLength: " << preset.LensSettings.MinFocalLength << "; \t MaxFocalLength: " << preset.LensSettings.MaxFocalLength;
```
Is storing the data in a string really the best way, makes it that much harder for users to extract and store data programmatically...
Upto you and the maintainers

ppueyor replied:
This is just informative. Users can use this method to extract all the preset lens available, so they can choose it with "setPresetLensSettings".

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+{
+    std::vector<std::string> vector;
+    TArray<FNamedFilmbackPreset> lens_presets = camera_->GetFilmbackPresets();
+    for (FNamedFilmbackPreset preset : lens_presets) {
```
`const FNamedFilmbackPreset& preset`, to avoid unnecessary copies, in other places also

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+    ss << "; \t Max Fstop: " << current_lens_params.MaxFStop;
+    ss << "\n\t Diaphragm Blade Count: " << current_lens_params.DiaphragmBladeCount;
+    ss << "\n\t Minimum focus distance: " << current_lens_params.MinimumFocusDistance;
+    std::string final_string(ss.str());
+
+    return final_string;
```
Can just directly return `ss.str()` here rather than extra local var which isn't gonna be used. In other places also

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+
+std::string APIPCamera::getLensSettings()
+{
+    std::vector<std::string> vector;
```
Unused var, in other places also

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+//CinemAirSim methods
+std::vector<std::string> APIPCamera::getPresetLensSettings()
+{
+    std::vector<std::string> vector;
```
Instead of `vector`, should use more descriptive var names, maybe `all_settings` or any other appropriate name

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+
+    std::ostringstream ss;
+
+    FString lens_preset_name = camera_->GetLensPresetName();
+    std::string name = (TCHAR_TO_UTF8(*lens_preset_name));
```
Try to add `const` to vars if they aren't modified, to avoid coding errors, and makes the intention clear, other places also

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+    return camera_->Filmback.SensorAspectRatio;
+}
+
+float APIPCamera::getFocalLength()
```
Such methods should also be marked as `const` to ensure and indicate to the reader they aren't modifying the state

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -605,3 +615,187 @@ void APIPCamera::onViewModeChanged(bool nodisplay)
+
+void APIPCamera::copyCameraSettingsToSceneCapture(UCameraComponent* src, USceneCaptureComponent2D* dst)
+{
+    if (src && dst) {
```
Maybe a log message in the `else` block?

ppueyor replied:
For now I haven't added anything, thus it is just for developers. If you have some suggestions of what we can log, then I can edit it.

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.h:
```diff
@@ -39,6 +43,28 @@ class AIRSIM_API APIPCamera : public ACameraActor
+    void enableFocusPlane(bool enable);
+    std::string getCurrentFieldOfView();
+
+    void copyCameraSettingsToAllSceneCapture(UCameraComponent* camera);
+    void copyCameraSettingsToSceneCapture(UCameraComponent* src, USceneCaptureComponent2D* dst);
```
These should be protected or private methods since they're meant to be used only by other member methods

rajat2004 reviewed (commented):
Some code improvement comments. Additionally, would be great if you could add some example scripts to test the various capabilities, and also some documentation.
Thanks!

Commit:
Merge branch 'microsoft:master' into master

Commit:
Changes requested in PR. Minor fixes, memory and naming conventions

ppueyor commented:
Hi @zimmy87 and @rajat2004 , I resolved all the changes requested, hope everything is correct now. 
There is some docu/demo work to do, but I don't have much time to do it right now. I will do it asap.

ppueyor commented:
> The new focal length looks good when I manually override the right camera, but unfortunately I'm not seeing this focal length applied to the default camera, so I still see the same narrow field of view with the latest iteration.

Does the "PostProcesing" thing resolves this?

zimmy87 commented:
Tested the latest iteration and it works well for me, so I am moving ahead with merging. It would be nice to add some extra documentation and examples, but I feel that can be handled in a separate PR.
